File: 98-validate-binary-search-tree/98-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool solve(TreeNode* root, long mi, long ma){
        if(!root) return true;
        return root->val>mi && root->val<ma && solve(root->left,mi,root->val) && solve(root->right,root->val,ma);
    }
    bool isValidBST(TreeNode* root) {
        return solve(root,LONG_MIN,LONG_MAX);
    }
};

File: 98-validate-binary-search-tree/98-validate-binary-search-tree.py
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def solve(root,maxi=inf,mini=-inf):
            if not root:return True
            cur = root.val
            return mini<cur<maxi and solve(root.left,min(maxi,cur),mini) and solve(root.right,maxi,max(mini,cur))        
        return solve(root)